examples/common/generate_callback.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Monitor rate of change of loss."""
+from typing import List
+
+import wandb
+from composer.core import Callback, State
+from composer.loggers import Logger, WandBLogger
+from composer.utils import dist, ensure_tuple
+
+
+class Generate(Callback):
+
+    def __init__(self, prompts: List[str], **kwargs):
+        self.prompts = prompts
+        self.generate_kwargs = kwargs
+
+    def eval_end(self, state: State, logger: Logger):
+        model = state.model
+        tokenizer = state.model.tokenizer
+        device = state.device
+
+        # stash the original original value of padding_side because generation requires left padding
+        original_padding_side = tokenizer.padding_side
+        tokenizer.padding_side = 'left'
+        if tokenizer.pad_token_id is None:
+            tokenizer.pad_token_id = tokenizer.eos_token_id
+        tokenized_input = tokenizer(self.prompts,
+                                    return_tensors='pt',
+                                    padding=True)
+
+        for k, v in tokenized_input.items():
+            tokenized_input[k] = device.tensor_to_device(v)
+
+        output_token_ids = model.model.generate(
+            input_ids=tokenized_input['input_ids'],
+            attention_mask=tokenized_input['attention_mask'],
+            synced_gpus=True,
+            **self.generate_kwargs,
+        )
+
+        if dist.get_global_rank() == 0:
+            wandb_artifact = wandb.Artifact('generate_samples_' +
+                                            str(wandb.run.id),
+                                            type='predictions')
+            for destination in ensure_tuple(logger.destinations):
+                if isinstance(destination, WandBLogger):
+                    rows = []
+                    for i in range(len(self.prompts)):
+                        prompt = self.prompts[i]
+                        output_tokens = output_token_ids[i][
+                            tokenized_input['input_ids'].shape[1]:]
+                        output_text = tokenizer.decode(output_tokens,
+                                                       skip_special_tokens=True)
+
+                        rows.append([prompt, output_text])
+
+                    text_table = wandb.Table(data=rows,
+                                             columns=['prompt', 'generation'])
+                    wandb_artifact.add(text_table, 'predictions')
+                    wandb.log_artifact(wandb_artifact)
```

Comment:
QQ, is this preferable to just logging the table via `wandb.log({"generate_samples" : text_table})`

Reply:
Artifacts always seem to add complexity... maybe we can also just log both if the reasoning is that that the Artifacts are easier to download?

Reply:
I started out with what you had...wandb does _not_ make it easy to view the evolution of this over the course of training. As far as I can tell the two options are

1. Log as table. The table in the UI will get overwritten  each time you write a new one, but it will show up in the main chart view (with an absurd amount of whitespace)
2. Log as artifact. The table will not show up in the main chart view, but if you click through like five menus, you have the ability to view all the generations over the course of training (in a pretty rough UI).

Do you have a different suggestions based on your wandb knowledge? Entirely possible I missed something.

Reply:
Should `self.wandb_artifact = wandb.Artifact(*args)` instantiation be done durring callback init?

Reply:
turns out no. once you send an artifact off to wandb, that object cannot be used anymore.
`ValueError: Can't add to finalized artifact.`

Reply:
Discussed offline, going to log both as table and artifact to get the best of both worlds. its still a pretty rough UI